### ISSUE TITLE
Please support PHP 7 by using the current php-mongodb extension

### ISSUE THREAD

**Issue by limikael** [LearningLocker/learninglocker#868]
**Version**
1.13.5

**Steps to reproduce the bug**
1. Install ubuntu 16.04 and php, this will give you PHP 7.
2. Follow the steps on: http://docs.learninglocker.net/installation/

**Expected behaviour**
I expected LearningLocker to be able to install.

**Actual behaviour**
It is breaking because LearningLocker uses the [legacy mongodb php driver](https://github.com/mongodb/mongo-php-driver-legacy/releases). This driver doesn't support php7, this is what happens when I try to install the driver using pecl (the pecl package for this driver is called `mongo`):

```
micke@S751:/var/www/html/learninglocker$ sudo pecl install mongo
WARNING: "pecl/mongo" is deprecated in favor of "channel:///mongodb"
pecl/mongo requires PHP (version >= 5.3.0, version <= 5.99.99), installed version is 7.0.8-0ubuntu0.16.04.2
No valid packages found
install failed
```

There is a new driver, this package is called `mongodb` and it installs fine, both when installing using pecl and when installing using `apt-get install php-mongodb`. However, LearningLocker cannot use this driver and this is the output I'm getting when I try to install LearningLocker using `composer`:

```
micke@S751:/var/www/html/learninglocker$ composer install --no-interaction --no-dev
Loading composer repositories with package information
Installing dependencies from lock file
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested PHP extension ext-mongo * is missing from your system. Install or enable PHP's mongo extension.
```

**Additional information**
OS: Ubuntu 16.04
Browser: N/A - I used the command line to install.

**Comment by ht2** [comment]
Hi @limikael,

Learning Locker does not support PHP 7, the compatibility goes way beyond just the Mongo driver. And would require an entire rewrite to the underlying application.

If you wish to contribute to a new version, then please consider forking the project and submitting pull requests, but as it stands there is no development plan to take the current version of Learning Locker into PHP7.

**Comment by limikael** [comment]
I see. Thanks anyway!

**Comment by andrewdisney** [comment]
I am curious what is not working in PHP7.  We are running PHP7 in our dev environment and learning locker seems to be running fine. Is there a list somewhere of items to be worked on?

**Comment by ht2** [comment]
Hey @andrewdisney, I'm curious.... Apart from a bit of change in syntax between versions (I believe changes to how Exceptions are handled are a problem moving between 5 and 7) the main problem is the Mongo driver compatbility: https://docs.mongodb.com/ecosystem/drivers/driver-compatibility-reference/#reference-compatibility-mongodb-php

Learning Locker uses the old PHP Mongo driver, which as far as I know has no compatibility with PHP 7. Can you confirm you are running vanilla Learning Locker on PHP 7 using this driver?

![image](https://cloud.githubusercontent.com/assets/1352590/21801736/d9fc00fe-d71a-11e6-8423-018dd56d475b.png)

**Comment by limikael** [comment]
That's what I wrote in the initial bug, that the problem was the mongo driver. However, you said at that point @ht2 that the main problem was actually _not_ the driver. You said "Learning Locker does not support PHP 7, the compatibility goes way beyond just the Mongo driver. And would require an entire rewrite to the underlying application". So what is the problem, is it the driver or something else?

**Comment by ht2** [comment]
I said it goes beyond just the Mongo driver... 

Laravel 4.2 does not work with PHP 7 (hence my mention of a rewrite to the underlying application) but to be honest we have not used any resources attempting to check that, but please feel free. Here is a potential useful resource as to blockers you might hit: https://laracasts.com/discuss/channels/laravel/laravel-42-and-php-7.

I'd suggest a good first step however would be to rewrite the repositories to use the newer Mongo driver (which is compatible with _both_ PHP 5.x and 7). That could lead onto an update to the core application to work with PHP 7 as well.

**Comment by ht2** [comment]
Looking here: https://github.com/laravel/framework/issues/11329 it seems that the newest version of 4.2 (4.2.20) has PHP 7 support. Is it possible that the old driver has also been (unofficially) updated for PHP 7 support too?

**Comment by limikael** [comment]
Ok I see... I'm not sure if I will have time to hack on this since we actually decided to create our own LRS and use that instead... It is called [wp-xapi-lrs](https://github.com/tunapanda/wp-xapi-lrs) and comes as a WordPress plugin, which makes it really convenient to use and install if you are using WordPress. If you are not using WordPress, most of it is implemented as a [library](https://github.com/limikael/minixapi/) which can be used without WordPress. The functionality is kind of limited at the moment, but at least it works to store statements and do basic queries. We would love some feedback @andrewdisney, so if you want, check it out!

**Comment by ht2** [comment]
No problem - it might be somethign we look at. It seems that this package (https://packagist.org/packages/mongodb/mongodb) "implements an API similar to that of the legacy PHP driver" over the top of the new driver, so it could be that by using that package, as well as ensuring v4.2.20 and removing the composer dependency to `ext-mongo` that it might well work in PHP7, @andrewdisney can confirm their setup I guess.

**Comment by andrewdisney** [comment]
We did have to add an adapter library the one we used was:
https://github.com/alcaeus/mongo-php-adapter

So we added "alcaeus/mongo-php-adapter": "^1.0" to our composer.json ran composer and the site seems to work fine. We are still testing though. We don't plan to roll out to our production environment for a few weeks maybe months if there are still compatibility issues.

We have other applications thought that require PHP 7. So that is why I was asking if there was some way we could help with anything not working in PHP 7. Just let us know.
